src/main/java/com/minecolonies/coremod/colony/buildings/inventory/ItemStackHandlerWithIndex.java

```diff
@@ -0,0 +1,333 @@
+    }
+
+    /**
+     * Sets the meta index and the reference value for it
+     */
+    public void setMetaIndex(ItemStorageIndex<Value> meta, Value metaIndexValue)
+    {
+        if (meta != null && this.meta != meta)
+        {
+            this.meta = meta;
+            this.metaIndexValue = metaIndexValue;
+
+            // Set the handlers content for the meta index
+            for (ItemStorage store : index.getIndexMap().keySet())
+            {
+                meta.addToIndex(store, metaIndexValue);
+            }
+        }
+    }
+
+    /**
+     * Clears the meta index reference and its content.
+     */
+    public void clearMetaIndex()
+    {
+        if (meta == null)
+        {
+            return;
+        }
+
+        // Remove stored items in the meta index
+        for (ItemStorage store : index.getIndexMap().keySet())
+        {
+            meta.removeFromIndex(store, metaIndexValue);
+        }
+
+        // Remove references
+        meta = null;
+        metaIndexValue = null;
+    }
+
+    @Override
+    public void setStackInSlot(int slot, ItemStack pstack)
+    {
+        if (!isValidSlot(slot))
+        {
+            return;
+        }
+
+        setStorageInSlot(slot, new ItemStorage(pstack));
+    }
+
+    /**
+     * Adds the storage to the index and the stack to the inventory.
+     *
+     * @param slot    Slot the item is added at
+     * @param storage ItemStorage to add
+     */
+    public void setStorageInSlot(Integer slot, ItemStorage storage)
+    {
+        if (!isValidSlot(slot))
+        {
+            return;
+        }
+
+        // Overwriting old stack
+        if (!ItemStackUtils.isEmpty(getStackInSlot(slot)))
+        {
+            ItemStorage store = new ItemStorage(getStackInSlot(slot));
+            index.removeFromIndex(store, slot);
+
+            if (meta != null)
+            {
+                meta.removeFromIndex(store, metaIndexValue);
+            }
+            usedSlots--;
+        }
+
+        // Some operations on forge side do not use extractitem cleanly, but instead use it to simulate and
+        // then set the slot to empty via this call, so gotta support setting empty stacks.
+        if (!ItemStackUtils.isEmpty(storage.getItemStack()))
+        {
+            // Adding stack to indexes
+            index.addToIndex(storage, slot);
+
+            if (meta != null)
+            {
+                meta.addToIndex(storage, metaIndexValue);
+            }
+            usedSlots++;
+        }
+        super.setStackInSlot(slot, storage.getItemStack());
+    }
+
+    @Override
+    @Nonnull
+    public ItemStack insertItem(int slot, @Nonnull ItemStack stack, boolean simulate)
+    {
+        if (!isValidSlot(slot) || ItemStackUtils.isEmpty(stack))
+        {
+            return stack;
+        }
+
+        int insert = getStackLimit(slot, stack);
+
+        if (getStackInSlot(slot) != ItemStack.EMPTY)
+        {
+            if (!getStackInSlot(slot).equals(stack))
```

Comment:
You need to check for mergability not for equality!
Merge if possible or return original stack.